Migrate Activity Database to Entity Framework Core

barronpm opened the pull request:
**Changes**
Migrates the activity database code over to EF Core. Changes internal application logic to use the new EF Core entities except for events, and adds a translation layer for the API and event systems.
All new projects treat warnings as errors. 

**Related Issues**
#42 
#1481 
#1895 
#2517 
#2518 
#2519 
#2520 
#2521 
#2522 

**Future Plans**

- The next step is to migrate user data to the new schema

- Once we are done creating the schema, we can remove the diagram files, and move the generated classes to `Jellyfin.Model`, as well as document them.

- We should fix the DI scopes so that `AddDbContextPool()` can be used instead of the current situation, where new `DbContext` instances are created every time the application connects to the database.

- When the API is rewritten, it should take advantage of the new EF Core entities so that the translation  layer can be removed.

sparky8251 commented:
Well, my initial tests went well. Even found a bug by supplying a very old database!

Since this data has to be ordered, the migration can take some time for larger activity tables. 6117 activities takes about ~5m to migrate.

Ideally most of the remaining data will not need ordering so it wont be slow, but even if it is the good news is that this will only have to happen about 2 more times.

No data loss, can even restart if the migration is failed. If there is a bug in a successful migration we can roll it back and try over with a fix, etc.

Very well made PR that should act as a solid foundation for future EF Core migrating. Hopefully some folks can review the code side soon :)

barronpm commented:
In relation to code review, any file that ends with `.g.cs` is auto generated and can be safely ignored

crobibero commented on Jellyfin.Data/Jellyfin.Data.csproj:
```diff
@@ -0,0 +1,46 @@
+<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <TargetFramework>netstandard2.0</TargetFramework>
```
```suggestion
    <TargetFramework>netstandard2.1</TargetFramework>
    <GenerateAssemblyInfo>false</GenerateAssemblyInfo>
    <GenerateDocumentationFile>true</GenerateDocumentationFile>
    <TreatWarningsAsErrors>true</TreatWarningsAsErrors>
```

barronpm replied:
`MediaBrowser.Model` targets 2.0 and 2.1, so either this would have to target both as well, or I could make `MediaBrowser.Model` only target 2.1

barronpm replied:
I updated both `MediaBrowser.Model` and `Jellyfin.Data` to netstandard 2.1, there were no problems in my testing.

crobibero commented on Jellyfin.Data/Jellyfin.Data.csproj:
```diff
@@ -0,0 +1,46 @@
+    <Folder Include="Enums\" />
+    <Folder Include="Structs\" />
+  </ItemGroup>
+
```
```suggestion
 <PropertyGroup Condition=" '$(Configuration)' == 'Debug' ">
    <CodeAnalysisRuleSet>../jellyfin.ruleset</CodeAnalysisRuleSet>
  </PropertyGroup>
```

barronpm replied:
Unfortunately, I can't enable `TreatWarningsAsErrors` in this project yet, but once this is merged, I'll make a PR that addresses this

crobibero replied:
Ok, sounds good

crobibero commented on Jellyfin.Server.Implementations/Jellyfin.Server.Implementations.csproj:
```diff
@@ -0,0 +1,17 @@
+<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <TargetFramework>netstandard2.1</TargetFramework>
```
```suggestion
    <TargetFramework>netstandard2.1</TargetFramework>
    <GenerateAssemblyInfo>false</GenerateAssemblyInfo>
    <GenerateDocumentationFile>true</GenerateDocumentationFile>
    <TreatWarningsAsErrors>true</TreatWarningsAsErrors>
```

crobibero commented on Jellyfin.Server.Implementations/Jellyfin.Server.Implementations.csproj:
```diff
@@ -0,0 +1,17 @@
+    <ProjectReference Include="..\MediaBrowser.Controller\MediaBrowser.Controller.csproj" />
+    <ProjectReference Include="..\MediaBrowser.Model\MediaBrowser.Model.csproj" />
+  </ItemGroup>
+
```
```suggestion
<PropertyGroup Condition=" '$(Configuration)' == 'Debug' ">
    <CodeAnalysisRuleSet>../jellyfin.ruleset</CodeAnalysisRuleSet>
  </PropertyGroup>
```

crobibero commented on Jellyfin.Server/Migrations/Routines/MigrateActivityLogDb.cs:
```diff
@@ -0,0 +1,101 @@
+{
+    public class MigrateActivityLogDb : IMigrationRoutine
+    {
+        public Guid Id => Guid.Parse("{3793eb59-bc8c-456c-8b9f-bd5a62a42978}");
```
```suggestion
        public Guid Id => Guid.Parse("3793eb59-bc8c-456c-8b9f-bd5a62a42978");
```

crobibero commented:
> 
> 
> In relation to code review, any file that ends with `.g.cs` and all the files in `Jellyfin.Data.Migrations` are auto generated and can be safely ignored

EraYaN commented:
So this should be merged after: #2293 I think otherwise we are going to see a horrible amount of weird merge issues with that.

barronpm commented:
This PR includes all of the changes from that PR so it is no longer necessary.

EraYaN commented:
Well but that PR will need changes (for example v2 of the extension generates nicer looking code), so this one should be rebased on that when it's in.

barronpm commented:
If the code needs to be regenerated, it should be done on this PR since I have added additional parts to the schema.

EraYaN commented:
Alright if we merge the other one first, then this one there should be no merge conflicts.

mark-monteiro commented:
Can we get a summary of what's been changed here (i.e. what functionality the activity database supports, what has changed in the schema, breaking changes, if any, how data migration is supported, etc). Maybe also an outline of some testing scenarios? This PR is 9000 lines long and I'm going in here blind without any context since I haven't closely followed the chat around this

mark-monteiro commented on Jellyfin.Server.Implementations/Activity/ActivityManager.cs:
```diff
@@ -0,0 +1,107 @@
+                ItemId = entry.ItemId,
+                UserId = entry.UserId,
+                Date = entry.DateCreated,
+                Severity = entry.LogSeverity
```
What about `UserPrimaryImageTag`?

barronpm replied:
As far as I can tell, that was completely unused. It was never stored in the database, and no code calls it. I don't think it should be, as that should be stored within the referenced User.

mark-monteiro replied:
Ah okay makes sense. Maybe let's just get rid of that property then? We might not want to change the object being returned from the API though, so maybe just add a `[Deprecated]` tag to it?

mark-monteiro commented on Jellyfin.Data.Migrations/JellyfinDbFactory.cs:
```diff
@@ -0,0 +1,21 @@
+
+namespace Jellyfin.Data.Migrations
+{
+    class JellyfinDbFactory : IDesignTimeDbContextFactory<JellyfinDb>
```
If I fix the null return below and run `dotnet ef migration add Foo` in this project directory I get the following error message, I wonder if the cause is that this needs to be in the same project as the DbContext?

> Your target project 'Jellyfin.Data.Migrations' doesn't match your migrations assembly 'Jellyfin.Data'. Either change your target project or change your migrations assembly.
Change your migrations assembly by using DbContextOptionsBuilder. E.g. options.UseSqlServer(connection, b => b.MigrationsAssembly("Jellyfin.Data.Migrations")). By default, the migrations assembly is the assembly containing the DbContext.
Change your target project to the migrations project by using the Package Manager Console's Default project drop-down list, or by executing "dotnet ef" from the directory containing the migrations project.

What was the reason for splitting up the projects this way? Normally I would put this and the migrations together with the dbContext and maybe split out the models into their own netstandard project, so this seems like a strange setup to me

barronpm replied:
So the original reason for splitting it into its own project was because Migrations require the underlying project to target .NET Core, not .NET Standard, but this is still a problem, as that applies to the project that contains the `DbContext` as well. I would like to migrate the projects to target .NET Core, but people in chat seemed wary of the idea.

barronpm replied:
To fix the above error, you would have to change the line `optionsBuilder.UseSqlite("./jellyfin.db");` to `optionsBuilder.UseSqlite("./jellyfin.db", x => x.MigrationsAssembly("Jellyfin.Data.Migrations"));`

mark-monteiro replied:
Ah okay great, as long as there is a way to make the `dotnet ef` tools work I'm fine with splitting them this way.

mark-monteiro commented on Jellyfin.Data.Migrations/JellyfinDbFactory.cs:
```diff
@@ -0,0 +1,21 @@
+            var optionsBuilder = new DbContextOptionsBuilder<JellyfinDb>();
+            optionsBuilder.UseSqlite("./jellyfin.db");
+
+            return null;
```
Looks like this isn't fully implemented yet?

barronpm replied:
Correct, I didn't think it was included in this PR, but the migration code is as of now unused. This code specifically would be used to create instances of `JellyfinDb` at design time, to be used in creating new migrations.

mark-monteiro commented on Jellyfin.Data.Migrations/Jellyfin.Data.Migrations.csproj:
```diff
@@ -0,0 +1,28 @@
+  <ItemGroup>
+    <ProjectReference Include="..\Jellyfin.Data\Jellyfin.Data.csproj" />
+  </ItemGroup>
+
```
Analyzers should be included for new projects

```suggestion
  <!-- Code analysers-->
  <ItemGroup Condition=" '$(Configuration)' == 'Debug' ">
    <PackageReference Include="Microsoft.CodeAnalysis.FxCopAnalyzers" Version="2.9.8" PrivateAssets="All" />
    <PackageReference Include="SerilogAnalyzer" Version="0.15.0" PrivateAssets="All" />
    <PackageReference Include="StyleCop.Analyzers" Version="1.1.118" PrivateAssets="All" />
    <PackageReference Include="SmartAnalyzers.MultithreadingAnalyzer" Version="1.1.31" PrivateAssets="All" />
  </ItemGroup>
```

mark-monteiro commented on Jellyfin.Data.Migrations/Jellyfin.Data.Migrations.csproj:
```diff
@@ -0,0 +1,28 @@
+  </PropertyGroup>
+
+  <PropertyGroup Condition="'$(Configuration)|$(Platform)'=='Debug|AnyCPU'">
+    <TreatWarningsAsErrors>true</TreatWarningsAsErrors>
```
```suggestion
    <TreatWarningsAsErrors>true</TreatWarningsAsErrors>
    <GenerateDocumentationFile>true</GenerateDocumentationFile>
```

mark-monteiro commented on Emby.Server.Implementations/Activity/ActivityLogEntryPoint.cs:
```diff
@@ -263,208 +271,253 @@ private static string GetPlaybackStoppedNotificationType(string mediaType)
 
             if (string.IsNullOrEmpty(session.UserName))
             {
+                name = string.Format(
+                    CultureInfo.InvariantCulture,
+                    _localization.GetLocalizedString("DeviceOfflineWithName"),
+                    session.DeviceName);
+
+                // Causing too much spam for now
```
Maybe add a `TODO:` comment here, but I don't think we want to add dead code
```suggestion
```

mark-monteiro commented on Emby.Server.Implementations/Activity/ActivityLogEntryPoint.cs:
```diff
@@ -263,208 +271,253 @@ private static string GetPlaybackStoppedNotificationType(string mediaType)
 
             if (string.IsNullOrEmpty(session.UserName))
             {
+                name = string.Format(
+                    CultureInfo.InvariantCulture,
+                    _localization.GetLocalizedString("DeviceOnlineWithName"),
+                    session.DeviceName);
+
+                // Causing too much spam for now
```
```suggestion
```

Same.

mark-monteiro commented on Emby.Server.Implementations/Activity/ActivityLogEntryPoint.cs:
```diff
@@ -495,22 +548,21 @@ private void OnTaskCompleted(object sender, TaskCompletionEventArgs e)
-        private void CreateLogEntry(ActivityLogEntry entry)
-            => _activityManager.Create(entry);
+        private async Task CreateLogEntry(ActivityLog entry)
+            => await _activityManager.CreateAsync(entry).ConfigureAwait(false);
```
This method seems really pointless, but I guess it was here before so okay 🤷

barronpm replied:
Yeah, I don't really know why that was there...

barronpm commented on Jellyfin.Server/Migrations/Routines/MigrateActivityLogDb.cs:
```diff
@@ -0,0 +1,101 @@
+                    }
+
+                    dbContext.ActivityLogs.Add(newEntry);
+                    dbContext.SaveChanges();
```
That would be what I would do normally, but this data needs to be inserted in order, and since EF Core provides no way to guarantee order during bulk inserts, this is what I did.

mark-monteiro replied:
Why do the records need to be inserted in-order? AFAIK insert order does not matter for SQL databases, since you always need to specify an `ORDER BY` clause to get sorted data

barronpm replied:
There is code that references the ID (Insertion order) of the records

barronpm commented on Jellyfin.Server.Implementations/JellyfinDbProvider.cs:
```diff
@@ -0,0 +1,35 @@
+namespace Jellyfin.Server.Implementations
+{
+    /// <summary>
+    /// Acts as an intermediate layer between the application's service provider and
+    /// the rest of the application to provide access to a connection pool to the
+    /// underlying database.
```
Fair enough. When I originally documented this the underlying `DbContext` service provider was using `AddDbContextPool`. This works by reusing instances within DI scopes, so I'm not sure if it is even possible to reimplement given that the same endpoint can be called from different threads simultaneously. This is definitely more accurate now

mark-monteiro replied:
Ah okay, definitely makes more sense in the context of `AddDbContextPool`.

I _think_ what happen to happen to get scoped requests working correctly is that over time is we need to switch over services to register them as 'scoped' instead of 'singleton'. That way a separate new instance of each service will be constructed for each request and can safely have `DbContext` injected directly via the constructor. A project for another day haha

barronpm replied:
From my understanding, scoping means to define separate sections of code where instances are reused within those sections. What you seem to be describing sounds more like a transient service to me, but I may be mistaken.

mark-monteiro replied:
I may also be mistaken but here is my understanding:

'Scoped' means that a new instance is created once per HTTP request. Generally the pattern is to register the DbContext and your controllers/services as scoped. This way, each request has it's own set of controllers/services and they all share the same `DbContext` _within that request scope_. So if you have two requests to the same controller in parallel, then two different instances of the controller and DbContext are created. This way you can share resources for the request, but also have thread safety, since each request runs on a single thread

barronpm commented on Jellyfin.Server.Implementations/Activity/ActivityManager.cs:
```diff
@@ -0,0 +1,107 @@
+
+            return new QueryResult<ActivityLogEntry>()
+            {
+                Items = list
```
Done. In the future I think we should change the return types of these methods, possibly to `IEnumerable` or `IList`

barronpm commented on Jellyfin.Server.Implementations/Activity/ActivityManager.cs:
```diff
@@ -0,0 +1,107 @@
+        }
+
+        /// <inheritdoc/>
+        public QueryResult<ActivityLogEntry> GetActivityLogEntries(DateTime? minDate, bool? hasUserId, int? startIndex, int? limit)
```
IMO this is definitely a good change and not out of scope for this PR. In the initial conversion I tried to match as closely as possible with the original methods, but since I am rewriting the class anyways, changing method signatures a bit more is not a problem.

barronpm commented on Emby.Server.Implementations/Activity/ActivityRepository.cs:
```diff
@@ -1,308 +0,0 @@
-            return info;
-        }
-    }
-}
```
🔥 🔥 🔥 🔥 🔥 🔥 🔥

barronpm commented on Jellyfin.Server/Migrations/Routines/MigrateActivityLogDb.cs:
```diff
@@ -0,0 +1,101 @@
+                ConnectionFlags.ReadOnly,
+                null))
+            {
+                logger.LogInformation("This may take a while, do not stop Jellyfin.");
```
Not sure if you would still want this changed, but this is called before each migration: `logger.LogInformation("Applying migration '{Name}'", migrationRoutine.Name);`

mark-monteiro replied:
Right, forgot about that. Maybe still something like "Migrating the database may take a while, do not stop Jellyfin." The use of "This"  in the original message just seems a bit too vague

mark-monteiro commented on Jellyfin.Server/Migrations/Routines/MigrateActivityLogDb.cs:
```diff
@@ -0,0 +1,101 @@
+
+                // Make sure that the database is empty in case of failed migration due to power outages, etc.
+                dbContext.Database.EnsureDeleted();
+                dbContext.Database.EnsureCreated();
```
I think this is incorrect, the [documentation for this method](https://docs.microsoft.com/en-us/dotnet/api/microsoft.entityframeworkcore.infrastructure.databasefacade.ensurecreated?view=efcore-3.1) says

> Note that this API does not use migrations to create the database. In addition, the database that is created cannot be later updated using migrations. If you are targeting a relational database and using migrations, you can use the DbContext.Database.Migrate() method to ensure the database is created and all migrations are applied. 

I think what we need to do is define an initial migration that creates the schema and then call `Database.Migrate()` somewhere in application startup (maybe in `Program.cs` or `ApplicationHost.InitializeServices()`)

barronpm replied:
Interesting. I didn't see the part about migrations not working on databases created using this method. That is a good catch and definitely needs to be fixed before this is merged

mark-monteiro replied:
Yeah, I'm very glad I caught this, that would have been quite a pain if we had initialized the database wrong for everyone on nightly hahah 😅

JustAMan commented:
@barronpm I have merged the base EFCore PR, could you please rebase your PR on master now? It should have way smaller diff hopefully...

EraYaN commented:
I have a rebase, let me push it

JustAMan commented:
Yay, 2k diff instead of 10k! 🎉

EraYaN commented:
I also mow notice the Guid vs. int keys... I'd rather stick with Int keys they are both faster and have much wider support. We can always shim the Guid to be the int key plus a server key and table key if we need to for the old API.

Thoughts?

JustAMan commented:
IMHO `int`s are preferable for at least internal indexes, as they should provide faster `JOIN`s and such. We can always add a guid field at the outermost `JOIN` if need be.

EraYaN commented:
Latest changes:

Remove all the full namespaces
Fluent API -> Attributes (+plus some fluent for the indexes)
Proper indentation (so hide whitespace changes during review)
Added proper cross backend RowVersioning (for concurrency checking)

barronpm commented:
Notes about the code for reviewers:
`Jellyfin.Server.Implementations.Migrations` is auto-generated, except for `JellyfinDesignTimeDbFactory`
Most of the changes are in `Jellyfin.Data`, and are just moving from the fluent API to attributes.

JustAMan commented:
> `Jellyfin.Server.Implementations.Migrations` is auto-generated

Where is source for that (mostly curious)?

EraYaN commented:
Probably `dotnet efcore` for the migrations.

EraYaN commented:
We might need to gate certain of these old -> new data migrations between or after certain EF core migrations. Or implement them as an EF Core migration if we can partially apply those.

barronpm commented on MediaBrowser.Api/Library/LibraryService.cs:
```diff
@@ -759,13 +759,14 @@ private void LogDownload(BaseItem item, User user, AuthorizationInfo auth)
             {
-                _activityManager.Create(new ActivityLogEntry
+                _activityManager.Create(new Jellyfin.Data.Entities.ActivityLog(
+                    string.Format(_localization.GetLocalizedString("UserDownloadingItemWithValues"), user.Name, item.Name),
+                    "UserDownloadingContent",
+                    auth.UserId,
+                    DateTime.UtcNow,
+                    LogLevel.Trace)
```
So this is actually something I was thinking about: The auto-generated code requires the non-null columns to be given in the constructor, while the old code left the responsibility to the caller to populate those fields. I think the latter approach looks much cleaner, at the cost of being slightly less safe.

sparky8251 replied:
Speaking up here, I feel higher safety is important. Ideally we never have to worry, but more safety means less burden on reviewers when community members send in PRs.

Anything we can do to stop randos from making a mess of things should be given a lot of weight imo.

mark-monteiro replied:
Yeah, I would also prefer to keep additional safety. As a compromise, we could add the non-required properties as optional parameters to the constructor

sparky8251 replied:
Yeah. Anything we can do to prevent extra review load and a new potential source of bugs from popping up.

Compiler should be our friend, the more we can ask it to do the better imo.

mark-monteiro commented:
Can we split this up into multiple PRs? IMO this should be split up into at least three separate reviews: 

1. Updating the generated model files to use attributes, other code cleanup etc (basically just commit 74871d571e032208243d7a17a321254afd26d1c2 from @EraYaN)
2. Restructuring the projects to add EF Core migration support
3. Migrating the ActivityDb (the original intent of this PR)

I'm open to alternatives, but as it stands this PR is such a monolith that it's kind of hard to approach. 

I would also propose a different approach to doing the migrations. For the first migration I think we should only add the tables needed for the `ActivityDb` to the database (create this migration as part of PR#3). I believe this can be easily done by commenting out all the other `DbSet<>` properties in the db context before creating the initial migration. Later on we can add additional migrations to add other tables as necessary, validating the schema as we go. I think this will be a cleaner approach and give us a better opportunity to review the database schema in smaller pieces and in the context of actually using the models. It will also allow us to make schema changes for the other tables in a cleaner way since we won't have to make `alter table` migrations

barronpm commented:
I mostly agree, but I think two PR's is enough and we can combine 2 and 3 without a huge diff. I agree with your idea of only adding the used parts to the database. I can do that when we pull the model file updates from this PR

barronpm commented:
Okay, I've split the other PR off, the changes are much smaller now. Don't worry about the diff count, a bit over 2000 of it is from moving the DbContext class, and another 1000 or so from the new entities I added to `Jellyfin.Data`

sparky8251 commented:
What's left to get this merged? I don't really see anything outstanding here.

crobibero commented on Jellyfin.Data/Entities/ActivityLog.cs:
```diff
@@ -0,0 +1,153 @@
+        /// <param name="userid"></param>
+        /// <param name="datecreated"></param>
+        /// <param name="logseverity"></param>
+        public ActivityLog(string name, string type, Guid userid, DateTime datecreated, Microsoft.Extensions.Logging.LogLevel logseverity)
```
```suggestion
        public ActivityLog(string name, string type, Guid userId, DateTime dateCreated, Microsoft.Extensions.Logging.LogLevel logSeverity)
```

crobibero commented on Jellyfin.Data/Entities/ActivityLog.cs:
```diff
@@ -0,0 +1,153 @@
+        /// <param name="userid"></param>
+        /// <param name="datecreated"></param>
+        /// <param name="logseverity"></param>
+        public static ActivityLog Create(string name, string type, Guid userid, DateTime datecreated, Microsoft.Extensions.Logging.LogLevel logseverity)
```
```suggestion
        public static ActivityLog Create(string name, string type, Guid userId, DateTime dateCreated, Microsoft.Extensions.Logging.LogLevel logSeverity)
```

crobibero commented on Jellyfin.Data/Entities/ActivityLog.cs:
```diff
@@ -0,0 +1,153 @@
+
+        /// <summary>
+        /// Required, Max length = 256
+        /// </summary>
+        [Required]
+        [MaxLength(256)]
+        [StringLength(256)]
+        public string Type { get; set; }
+
+        /// <summary>
+        /// Required
+        /// </summary>
+        [Required]
+        public Guid UserId { get; set; }
+
+        /// <summary>
+        /// Max length = 256
+        /// </summary>
+        [MaxLength(256)]
+        [StringLength(256)]
+        public string ItemId { get; set; }
+
+        /// <summary>
+        /// Required
+        /// </summary>
+        [Required]
+        public DateTime DateCreated { get; set; }
+
+        /// <summary>
+        /// Required
+        /// </summary>
+        [Required]
+        public Microsoft.Extensions.Logging.LogLevel LogSeverity { get; set; }
+
+        /// <summary>
+        /// Required, ConcurrenyToken
+        /// </summary>
+        [ConcurrencyCheck]
+        [Required]
+        public uint RowVersion { get; set; }
```
Properties should have actual documentation

barronpm replied:
I was planning on documenting all of the new entity classes in a separate PR, but if you think this is a blocker I can do it here

crobibero replied:
Making a docs pass works for me!

JustAMan commented on MediaBrowser.Api/System/ActivityLogService.cs:
```diff
@@ -53,7 +53,7 @@ public object Get(GetActivityLogs request)
                 DateTime.Parse(request.MinDate, null, DateTimeStyles.RoundtripKind).ToUniversalTime();
 
-            var result = _activityManager.GetActivityLogEntries(minDate, request.HasUserId, request.StartIndex, request.Limit);
+            var result = _activityManager.GetPagedResult(request.StartIndex, request.Limit);
```
why minDate is not used?

barronpm replied:
🤷 Looks like that's been there for 6 years

JustAMan commented on MediaBrowser.sln:
```diff
@@ -1,6 +1,6 @@
 Microsoft Visual Studio Solution File, Format Version 12.00
-# Visual Studio 15
-VisualStudioVersion = 15.0.26730.3
+# Visual Studio Version 16
+VisualStudioVersion = 16.0.30011.22
```
why we need this?

barronpm replied:
We don't, fixing

barronpm commented on Jellyfin.Server/Migrations/Routines/MigrateActivityLogDb.cs:
```diff
@@ -0,0 +1,120 @@
+                dbContext.ActivityLogs.RemoveRange(dbContext.ActivityLogs);
+                dbContext.SaveChanges();
+                // Reset the autoincrement counter
+                dbContext.Database.ExecuteSqlRaw("UPDATE sqlite_sequence SET seq = 0 WHERE name = 'ActivityLog';");
```
Currently AFAIK that is the only sqlite-specific code, and since it is run on the old db files, it is irrelevant to any future providers we add

JustAMan replied:
Not quite, this line I highlighted works on new DB. So we migrate _from_ sqlite _to_ something EFCore... Though this is fixable later for sure.

barronpm commented on Emby.Server.Implementations/Activity/ActivityLogEntryPoint.cs:
```diff
@@ -558,15 +588,15 @@ private static string ToUserFriendlyString(TimeSpan span)
             {
                 int years = days / DaysInYear;
                 values.Add(CreateValueString(years, "year"));
-                days %= DaysInYear;
+                days = days % DaysInYear;
```
If I had to guess, I fixed that in a cleanup pr I did a while back , but I forked this off of master before it was merged

barronpm commented on Jellyfin.Server.Implementations/Migrations/20200502231229_InitialSchema.Designer.cs:
```diff
@@ -0,0 +1,73 @@
+#pragma warning disable CS1591
```
Everything auto-generated now is from the EF Core migration tool

JustAMan replied:
can we add a `README.md` file inside that migrations directory describing how to make those?

barronpm commented on Jellyfin.Server.Implementations/Migrations/20200502231229_InitialSchema.cs:
```diff
@@ -0,0 +1,46 @@
+                columns: table => new
+                {
+                    Id = table.Column<int>(nullable: false)
+                        .Annotation("Sqlite:Autoincrement", true),
```
Yes, AFAIK, we would just create another migration for the new provider

barronpm commented on Emby.Server.Implementations/Activity/ActivityLogEntryPoint.cs:
```diff
@@ -272,110 +280,120 @@ private void OnSessionEnded(object sender, SessionEventArgs e)
                 ShortOverview = string.Format(
                     CultureInfo.InvariantCulture,
                     _localization.GetLocalizedString("LabelIpAddressValue"),
                     e.Argument.SessionInfo.RemoteEndPoint),
```
I'm not opposed to that, but I feel that it is out of scope for this PR

JustAMan replied:
I agree, was just thinking out loud

barronpm commented on Emby.Server.Implementations/Devices/DeviceManager.cs:
```diff
@@ -416,27 +275,9 @@ public class DeviceManagerEntryPoint : IServerEntryPoint
         }
 
-        public async Task RunAsync()
+        public Task RunAsync()
```
Since this does nothing, is it safe to remove the parent class `DeviceManagerEntryPoint` ?

JustAMan replied:
I think so. However I would really like this in a separate PR if possible, to limit the scope of single PR to an atomic set of changes. However, seeing you've already dropped other instances of camera upload, we might as well finish it here...

JustAMan commented on Jellyfin.Server/Migrations/Routines/MigrateActivityLogDb.cs:
```diff
@@ -56,9 +80,11 @@ public void Perform()
+                        entry[6].SQLiteType == SQLiteType.Null ? Guid.Empty : Guid.Parse(entry[6].ToString()))
+                    {
+                        DateCreated = entry[7].ReadDateTime(),
+                        LogSeverity = logLevelDictionary[entry[8].ToString()]
```
we should handle a case when `entry[8].ToString()` is not in a dictionary, i.e. make it default to `Trace` if nothing fits.

JustAMan commented on Jellyfin.Server/Migrations/Routines/MigrateActivityLogDb.cs:
```diff
@@ -75,6 +101,8 @@ public void Perform()
                     }
 
+                    // Since code references the Id of the entries, this needs to be inserted in order.
+                    // In order to do that, this is needed because EF Core doesn't provide a way to guarantee ordering for bulk inserts.
```
```suggestion
                    // In order to do that, we insert one by one because EF Core doesn't provide a way to guarantee ordering for bulk inserts.
```

mark-monteiro commented on Jellyfin.Data/Entities/ActivityLog.cs:
```diff
@@ -0,0 +1,150 @@
+         *************************************************************************/
+
+        /// <summary>
+        /// Identity, Indexed, Required
```
These can be specified here, but they should come after the standard summary text enforced by our stylecop rules. This applied to all the other properties below as well.
```suggestion
        /// Gets or set the id. Identity, Indexed, Required.
```

barronpm replied:
My original intention was to document the generated entities in a separate PR, but if you feel it is a blocker I can do it here

mark-monteiro replied:
Not all of the entities, just this one. This file is added in this PR so might as well document it here as well.

mark-monteiro commented on Jellyfin.Data/Entities/ActivityLog.cs:
```diff
@@ -0,0 +1,150 @@
+
+namespace Jellyfin.Data.Entities
+{
+    public partial class ActivityLog
```
```suggestion
    public partial class ActivityLog : ISavingChanges
```

mark-monteiro commented on Jellyfin.Data/Entities/ActivityLog.cs:
```diff
@@ -0,0 +1,150 @@
+        [Required]
+        public uint RowVersion { get; set; }
+
+        public void OnSavingChanges()
```
```suggestion
        /// <inheritdoc/>
        public void OnSavingChanges()
```

mark-monteiro commented on Emby.Server.Implementations/Emby.Server.Implementations.csproj:
```diff
@@ -50,7 +51,7 @@
   </ItemGroup>
 
   <PropertyGroup>
-    <TargetFramework>netstandard2.1</TargetFramework>
+    <TargetFramework>netcoreapp3.1</TargetFramework>
```
I don't think this is necessary, because this project doesn't need a reference to `Jellyfin.Server.Implementations`. The new services registered in `ApplicationHost` can be moved to `CoreAppHost` and then the reference to `Jellyfin.Server.Implementations` isn't necessary and this project can stay at `netstandard2.1`

mark-monteiro commented on Emby.Server.Implementations/ApplicationHost.cs:
```diff
@@ -545,6 +547,13 @@ protected virtual void RegisterServices(IServiceCollection serviceCollection)
+                options => options.UseSqlite($"Filename={Path.Combine(ApplicationPaths.DataPath, "jellyfin.db")}"),
+                ServiceLifetime.Transient);
+
+            serviceCollection.AddSingleton<JellyfinDbProvider>();
```
I think these service registrations should go into `CoreAppHost` instead

mark-monteiro commented on Jellyfin.Server/Jellyfin.Server.csproj:
```diff
@@ -41,6 +44,10 @@
+    <PackageReference Include="Microsoft.EntityFrameworkCore.Design" Version="3.1.3">
+      <PrivateAssets>all</PrivateAssets>
+      <IncludeAssets>runtime; build; native; contentfiles; analyzers; buildtransitive</IncludeAssets>
+    </PackageReference>
```
AFAIK this should be in the `Jellyfin.Server.Implementations` project since it is the project containing the `DbContext` and migrations

mark-monteiro commented on Jellyfin.Server/Jellyfin.Server.csproj:
```diff
@@ -13,6 +13,9 @@
     <Nullable>enable</Nullable>
+
+    <!-- Used for generating migrations for EF Core -->
+    <GenerateRuntimeConfigurationFiles>True</GenerateRuntimeConfigurationFiles>
```
Should this be in the `Jellyfin.Server.Implementations` project?

mark-monteiro commented on Jellyfin.Server.Implementations/Migrations/JellyfinDbModelSnapshot.cs:
```diff
@@ -0,0 +1,68 @@
+
+                    b.HasKey("Id");
+
+                    b.ToTable("ActivityLog");
```
```suggestion
                    b.ToTable("ActivityLogs");
```

mark-monteiro replied:
It looks like the migration was created with the wrong table name, it will have to be regenerated.

mark-monteiro commented on Jellyfin.Server.Implementations/JellyfinDbProvider.cs:
```diff
@@ -0,0 +1,33 @@
+        /// <returns>The newly created context.</returns>
+        public JellyfinDb CreateContext()
+        {
+            return _serviceProvider.GetService<JellyfinDb>();
```
```suggestion
            return _serviceProvider.GetRequiredService<JellyfinDb>();
```

mark-monteiro commented on Emby.Server.Implementations/Devices/DeviceManager.cs:
```diff
@@ -399,102 +209,4 @@ private static bool CanAccessDevice(UserPolicy policy, string id)
-
-        public async Task RunAsync()
-        {
-            if (!_config.Configuration.CameraUploadUpgraded && _config.Configuration.IsStartupWizardCompleted)
```
The `CameraUploadUpgraded` property can be deleted now

mark-monteiro commented on Emby.Server.Implementations/Devices/DeviceManager.cs:
```diff
@@ -194,172 +170,6 @@ private string GetDevicePath(string id)
-
-            if (string.IsNullOrWhiteSpace(name))
-            {
-                name = _localizationManager.GetLocalizedString("HeaderCameraUploads");
```
I think the "HeaderCameraUploads" translation string can be removed from the translation files now.

barronpm replied:
I think that should be left to a separate pr. Removing everything I have so far is already pretty far out of scope, and I think that would be enough to justify its own pr

mark-monteiro replied:
Sounds good, I would just open a new issue to make sure it is tracked and not forgotten about (unless you're going to open a separate PR yourself immediately instead)

mark-monteiro commented on Emby.Server.Implementations/Devices/DeviceManager.cs:
```diff
@@ -194,172 +170,6 @@ private string GetDevicePath(string id)
-        private Tuple<string, string, string> GetUploadPath(DeviceInfo device)
-        {
-            var config = _config.GetUploadOptions();
-            var path = config.CameraUploadPath;
```
Looks like `MediaBrowser.Model.Devices.DevicesOptions` can be remove now as well

barronpm replied:
DeviceOptions still has usages throught the code. If it can be removed, I think it should be done in its own PR

mark-monteiro replied:
I was referring to Device**s**Options, with an s. It's not used any more.

cvium commented on MediaBrowser.Api/Library/LibraryService.cs:
```diff
@@ -759,13 +759,12 @@ private void LogDownload(BaseItem item, User user, AuthorizationInfo auth)
             try
             {
-                _activityManager.Create(new ActivityLogEntry
+                _activityManager.Create(new Jellyfin.Data.Entities.ActivityLog(
```
I'm not particularly thrilled about using DB entities directly in the API layer. Imo the DB should be in a separate project that exposes service classes and DTO models.

@Bond-009 @mark-monteiro @EraYaN thoughts?

mark-monteiro replied:
@cvium Using DTO models would be my preference too, afaik that pattern is still widely used. That said, it might be better to handle in a follow-up PR, since the original code was also not using DTOs and the scope of this PR is already pretty large IMO

JustAMan replied:
I agree with Mark here, let's introduce middle-objects in a later PR. I also agree that we need a layer so we can decouple parts more, but this PR is already huge.

barronpm replied:
🙏

cvium replied:
As long as it is being handled. Preferably sooner rather than later.

barronpm commented on Jellyfin.Data/Entities/ActivityLog.cs:
```diff
@@ -0,0 +1,150 @@
+
+namespace Jellyfin.Data.Entities
+{
+    public partial class ActivityLog
```
@EraYaN was in favor of keeping them partial, but I agree that it isn't necessary

mark-monteiro replied:
@EraYaN Can you expand on why you wanted to keep these classes as `partial`?

mark-monteiro replied:
On second thought, we can deal with this in a separate PR

barronpm commented on Jellyfin.Data/Entities/ActivityLog.cs:
```diff
@@ -0,0 +1,150 @@
+        /// <param name="userId"></param>
+        /// <param name="datecreated"></param>
+        /// <param name="logSeverity"></param>
+        public ActivityLog(string name, string type, Guid userId)
```
If we want to avoid mixing arguments and property initialization we would also have to add optional parameters for Overview and ShortOverview

mark-monteiro replied:
That would probably be my preference, but maybe that will be too verbose for classes with more properties? I'd be interested to see what others in @jellyfin/backend think

barronpm commented on Jellyfin.Server.Implementations/Activity/ActivityManager.cs:
```diff
@@ -0,0 +1,103 @@
+            }
+
+            // This converts the objects from the new database model to the old for compatibility with the existing API.
+            var list = result.Select(entry => ConvertToOldModel(entry)).ToList();
```
This doesn't compile, that syntax only works if `ConvertToOldModel` is a method in `entry`'s type

mark-monteiro replied:
I already tested this locally and it compiles fine

> that syntax only works if ConvertToOldModel is a method in entry's type

AFAIK in that scenario you _wouldn't_ be able to do this?

barronpm replied:
You're right, was scatterbrained and made a typo 🤦

barronpm commented on Jellyfin.Server.Implementations/Activity/ActivityManager.cs:
```diff
@@ -0,0 +1,103 @@
+
+            if (startIndex.HasValue)
+            {
+                result = result.Where(entry => entry.Id >= startIndex.Value);
```
This is incompatible with calling `OrderByDescending` always, as that would skip the latest entries, not the oldest

mark-monteiro replied:
If you were using `OrderByDescending` then I think skipping the oldest entries would be the desired behavior. For example, if I want page 2 of results, sorted by newest, then the first page of results (i.e. newest) should be skipped

barronpm commented on Jellyfin.Server.Implementations/JellyfinDbProvider.cs:
```diff
@@ -0,0 +1,33 @@
+        public JellyfinDbProvider(IServiceProvider serviceProvider)
+        {
+            _serviceProvider = serviceProvider;
+            serviceProvider.GetService<JellyfinDb>().Database.Migrate();
```
I think we should have users who want to do manual migrations (something I don't think we should recommend, as they would have to manually perform operations EF Core does behind the scenes) perform them themselves. If they just want to make backups, they can do that before they start the new version for the first time (Something we recommend anyway)

JustAMan replied:
Well, we can at least add that "automatic migration" flag and plain refuse to start if it's not set... but we should tell the user how to force-run the migrations with it.

The point is, people might not be doing backups on each update, but they would if the update makes the server to stop booting.

PrplHaz4 replied:
Also, don't underestimate the number of people running Jellyfin on Docker with Watchtower that are getting the new version automatically the minute the new release is tagged.

barronpm replied:
I think all of the reviewers know this, but I think it's worth mentioning that the migration doesn't modify the original data, except to rename the file, in fact, the connection to the database is flagged as read only, so it would take some catastrophic failures to cause any data loss. On top of that, I feel like requiring an argument to automatically perform migrations is just delaying the problem until the second migration, when it is performed automatically for all of the users who set and forget the argument. Overall, I think it's too much work for little to no benefit

JustAMan replied:
Ohm, right, old DB is retained... I'm fine with it then.

mark-monteiro replied:
> requiring an argument to automatically perform migrations is just delaying the problem until the second migration, when it is performed automatically for all of the users who set and forget the argument

The point of the feature is that there will be users who will not set that flag and want to perform the migrations manually. Like I mentioned originally, there are two benefits:

1. Users can have control over how/when their database is migrated
2. The Jellyfin application does not need to have schema change permission on the database

IMO I think the benefits outweigh the cost here, since the only cost being the code suggestion in my comment and adding a single property to the server config.

That being said, I suppose this migration is relatively safe, so I would be okay to put it in a future PR if you don't want to add it here

barronpm replied:
Sounds good, it can be in a separate PR

barronpm commented on Jellyfin.Server.Implementations/Activity/ActivityManager.cs:
```diff
@@ -0,0 +1,103 @@
+
+        /// <inheritdoc/>
+        public QueryResult<ActivityLogEntry> GetPagedResult(
+            Func<IQueryable<ActivityLog>, IEnumerable<ActivityLog>> func,
```
When designing this signature, I was trying to minimize how much of the implementation is exposed to the caller. Is there any benefit to using `DbSet` over `IQueryable`?

mark-monteiro replied:
Yeah sorry, you're right `IQueryable` is better than `DbSet`. I think the second one should still be `IQueryable` instead of `IEnumerable` though

JustAMan commented on Jellyfin.Server.Implementations/Activity/ActivityManager.cs:
```diff
@@ -50,31 +50,31 @@ public async Task CreateAsync(ActivityLog entry)
             using var dbContext = _provider.CreateContext();
 
-            var result = func.Invoke(dbContext.ActivityLogs).AsQueryable();
+            var query = func(dbContext.ActivityLogs).OrderByDescending(entry => entry.DateCreated).AsQueryable();
```
why ordering is forced?

barronpm replied:
@mark-monteiro said that was desired behavior, and I double-checked and that is what the behavior was before

JustAMan commented on Jellyfin.Server/Jellyfin.Server.csproj:
```diff
@@ -13,9 +13,6 @@
     <Nullable>enable</Nullable>
-
-    <!-- Used for generating migrations for EF Core -->
-    <GenerateRuntimeConfigurationFiles>True</GenerateRuntimeConfigurationFiles>
```
why is it removed?

barronpm replied:
Apparently it's not needed, I added it because I saw something about that online and things started working, but then I removed it and everything still works, so 🤷. Also the Jellyfin.Server project isn't even where it would be added if it was necessary, so it's a mystery why adding that ever fixed anything

mark-monteiro commented on Jellyfin.Data/Jellyfin.Data.csproj:
```diff
@@ -1,12 +1,26 @@
 
   <PropertyGroup>
-    <TargetFramework>netstandard2.0</TargetFramework>
+    <TargetFrameworks>netstandard2.0;netstandard2.1</TargetFrameworks>
```
What's the reason to keep both of these? I'm not really familiar with the targeting setup for the solution, so this is more of a question

barronpm replied:
So @Bond-009 said that we need this project to target netstandard 2.0, and the rest of the project targets 2.1, so I'm taking the same approach as `MediaBrowse.Model` and targeting both

mark-monteiro replied:
Ah okay, yeah makes sense to match what is in `MediaBrowser.Model`

mark-monteiro commented:
Also leaving a comment here to track which issues we've decided to put into separate PRs so they don't get lost:
- update activity database to use DTOs isntead of returning entities directly (https://github.com/jellyfin/jellyfin/pull/2970#discussion_r424909191)
- remove remaining camera upload code (i.e. translation string "HeaderCameraUploads") (https://github.com/jellyfin/jellyfin/pull/2970#discussion_r424883259)
- Add config option to perform database migrations manually (https://github.com/jellyfin/jellyfin/pull/2970#discussion_r424863646)
- Consider adding params for optional proprties to entity constructors (https://github.com/jellyfin/jellyfin/pull/2970#discussion_r424871196)
- Consider removing `partial` from entity classes (https://github.com/jellyfin/jellyfin/pull/2970#discussion_r424870084)

mark-monteiro commented on Jellyfin.Server/Migrations/Routines/MigrateActivityLogDb.cs:
```diff
@@ -0,0 +1,125 @@
+                dbContext.Database.ExecuteSqlRaw("UPDATE sqlite_sequence SET seq = 0 WHERE name = 'ActivityLog';");
+                dbContext.SaveChanges();
+
+                foreach (var entry in queryResult)
+                {
+                    if (!logLevelDictionary.TryGetValue(entry[8].ToString(), out var severity))
+                    {
+                        severity = LogLevel.Trace;
+                    }
+
+                    var newEntry = new ActivityLog(
+                        entry[1].ToString(),
+                        entry[4].ToString(),
+                        entry[6].SQLiteType == SQLiteType.Null ? Guid.Empty : Guid.Parse(entry[6].ToString()))
+                    {
+                        DateCreated = entry[7].ReadDateTime(),
+                        LogSeverity = severity
+                    };
+
+                    if (entry[2].SQLiteType != SQLiteType.Null)
+                    {
+                        newEntry.Overview = entry[2].ToString();
+                    }
+
+                    if (entry[3].SQLiteType != SQLiteType.Null)
+                    {
+                        newEntry.ShortOverview = entry[3].ToString();
+                    }
+
+                    if (entry[5].SQLiteType != SQLiteType.Null)
+                    {
+                        newEntry.ItemId = entry[5].ToString();
+                    }
+
+                    dbContext.ActivityLogs.Add(newEntry);
+                }
```
We can also get another performance boost by using `AddRange` instead of calling `Add` within the loop. (NOTE: you'll need to add `using System.Linq;` above to apply this change). The diff on GitHub is long, but I just changed the foreach loop in the first line t use `Select` instead and also last few lines lines to use `AddRange` instead of `Add`

```suggestion
                var newEntries = queryResult.Select(entry =>
                {
                    if (!logLevelDictionary.TryGetValue(entry[8].ToString(), out var severity))
                    {
                        severity = LogLevel.Trace;
                    }

                    var newEntry = new ActivityLog(
                        entry[1].ToString(),
                        entry[4].ToString(),
                        entry[6].SQLiteType == SQLiteType.Null ? Guid.Empty : Guid.Parse(entry[6].ToString()))
                    {
                        DateCreated = entry[7].ReadDateTime(),
                        LogSeverity = severity
                    };

                    if (entry[2].SQLiteType != SQLiteType.Null)
                    {
                        newEntry.Overview = entry[2].ToString();
                    }

                    if (entry[3].SQLiteType != SQLiteType.Null)
                    {
                        newEntry.ShortOverview = entry[3].ToString();
                    }

                    if (entry[5].SQLiteType != SQLiteType.Null)
                    {
                        newEntry.ItemId = entry[5].ToString();
                    }

                    return newEntry;
                });
                
                dbContext.ActivityLogs.AddRange(newEntries);
```

mark-monteiro commented on Jellyfin.Server/Migrations/Routines/MigrateActivityLogDb.cs:
```diff
@@ -0,0 +1,125 @@
+            catch (IOException e)
+            {
+                _logger.LogError(e, "Error renaming legacy activity log database to 'activitylog.db.old'");
+            }
```
We should probably move the journal file as well so it is picked up in case the user opens the renamed old database

```suggestion
            }
            
            try
            {
                var journalPath = Path.Combine(dataPath, DbFilename + "-journal");
                if (File.Exists(journalPath))
                {
                    File.Move(journalPath, Path.Combine(dataPath, DbFilename + ".old-journal"));
                }
            }
            catch (IOException e)
            {
                _logger.LogError(e, "Error renaming legacy activity log database journal to 'activitylog.db.old-journal'");
            }
```

mark-monteiro commented on Jellyfin.Server/Migrations/Routines/MigrateActivityLogDb.cs:
```diff
@@ -0,0 +1,133 @@
+
+                    if (entry[5].SQLiteType != SQLiteType.Null)
+                    {
+                                        newEntry.ItemId = entry[5].ToString();
```
Some extra indentation got added here by mistake
